crates/ty_python_semantic/src/types/generics.rs

```diff
@@ -103,28 +112,57 @@ pub(super) fn walk_generic_context<'db, V: super::visitor::TypeVisitor<'db> + ?S
     visitor: &V,
 ) {
     for bound_typevar in context.variables(db) {
-        visitor.visit_bound_type_var_type(db, *bound_typevar);
+        visitor.visit_bound_type_var_type(db, bound_typevar);
     }
 }
 
 // The Salsa heap is tracked separately.
 impl get_size2::GetSize for GenericContext<'_> {}
 
 impl<'db> GenericContext<'db> {
+    pub(crate) fn variables(
+        self,
+        db: &'db dyn Db,
+    ) -> impl ExactSizeIterator<Item = BoundTypeVarInstance<'db>> {
+        match self.inner(db) {
+            GenericContextInner::Tuple { single_typevar } => {
+                Either::Left(std::iter::once(*single_typevar))
+            }
+            GenericContextInner::NonTuple(variables) => Either::Right(variables.iter().copied()),
+        }
+    }
+
     /// Creates a generic context from a list of PEP-695 type parameters.
     pub(crate) fn from_type_params(
         db: &'db dyn Db,
         index: &'db SemanticIndex<'db>,
         binding_context: Definition<'db>,
         type_params_node: &ast::TypeParams,
+        known_class: Option<KnownClass>,
     ) -> Self {
         let variables: FxOrderSet<_> = type_params_node
             .iter()
             .filter_map(|type_param| {
                 Self::variable_from_type_param(db, index, binding_context, type_param)
             })
             .collect();
-        Self::new(db, variables)
+
+        match known_class {
+            Some(KnownClass::Tuple) => {
+                assert_eq!(
+                    variables.len(),
+                    1,
+                    "Tuple should always have exactly one typevar"
+                );
+                Self::new(
+                    db,
+                    GenericContextInner::Tuple {
+                        single_typevar: variables[0],
+                    },
+                )
```

Comment:
This seems like dead code (assuming no custom typeshed), because this method is only used for PEP 695 type params, and the typeshed definition of `tuple` doesn't use PEP 695. Replacing this with `unreachable!(...)` doesn't fail any tests. Not sure how much we should account for custom typeshed with totally different `tuple` definition; I feel like we're already doing a lot of special-casing based on assumptions about the typeshed `tuple` (like, for instance, that it is generic over a single typevar, not a `TypeVarTuple`).

Reply:
Yeah, Doug and I thought it was good to future-proof the code here nonetheless, since one day typeshed will likely switch to the new syntax everywhere. Can add a comment.